pyo3-derive-backend/src/py_class.rs

```diff
@@ -133,9 +152,44 @@ fn impl_class(
         quote! {0}
     };
 
+    // If there are no type arguments, simply use a static variable.
+    let type_object_impl = if generics.params.is_empty() {
+        quote! {
+            static mut TYPE_OBJECT: ::pyo3::ffi::PyTypeObject = ::pyo3::ffi::PyTypeObject_INIT;
+            &mut TYPE_OBJECT
+        }
+    }
+    // If there are t-args, lacking support for t-arg dependent statics in Rust,
+    // we fall back to using a global hash map for resolving the correct type
+    // object using TypeId.
+    else {
+        quote! {
+            use std::collections::hash_map::Entry;
+            let mut map = ::pyo3::typeob::PY_TYPE_OBJ_MAP.lock().unwrap();
+            let obj = match map.entry(std::any::TypeId::of::<#cls #ty_generics>()) {
+                Entry::Occupied(o) => o.into_mut(),
+                Entry::Vacant(v) => v.insert(::pyo3::ffi::PyTypeObject_INIT),
+            };
+
+            // Erase ownership info using raw pointer
```

Comment:
What do you mean

Reply:
I think there's some possible ways..
1. LinkedList
2. HashMap<TypeId, Box<TypeObject>>
3. HashMap<TypeId, PyObject> (allocates objects in python heap)

And still not sure what is the best :thinking:

Reply:
I feel current `type_object() -> &' static mut ..` interface doesn't fit 'Typeobject in Python heap' cases...
But to handle heap objects manually in rust side makes me suspicious

Reply:
You can re-export stuff (e.g. the complete lazy_static crate) using `#[doc(hidden)]`. We're already using that in some places

Reply:
And **allocates Type objects every time in python heap** is a also possible solution.
`type_object()` doesn't called so frequently.

Reply:
I'm not sure about that, because we must make sure that objects of the same type are recognized in python as such, i.e. `type(SomeRustType(1)) == type(SomeRustType(2))`

Reply:
Ah, I see. Please forget what I said :)